### PR TITLE
 ENH: add support for NEP 35 constructors (`array`, `asarray`, `asanyarray`, `ascontiguousarray` and `asfortranarray`) with like=Quantity(...))

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description

ref #17124, builds on top of #17120

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/units/quantity_helper/function_helpers.py [hunk 1]
```diff
@@ -556,7 +557,57 @@ def full_impl(shape, fill_value, dtype=None, order='C', *, device=None):
...
+if NUMPY_LT_2_0:
+    @dispatched_function
+    def asarray(a, dtype=None, order=None, *, copy=None):
+        return asarray_impl(a, dtype, order, copy=copy)
+else:
+    @dispatched_function
+    def asarray(a, dtype=None, order=None, *, device=None, copy=None):
+        return asarray_impl(a, dtype, order, device=device, copy=copy)
+
+def asarray_impl(a, dtype=None, order=None, *, device=None, copy=None):
```
Comment (mhvk) [inline astropy/units/quantity_helper/function_helpers.py:575]: Like my comment for `np.zeros`, etc., I think all these functions could share a common implementation if one passed along `function=np.asarray`, etc.

It is a pity that for these the wrapping with `function_helper` does not work. Really, the `super` call should work - I guess this is a bug on the numpy side?

**Review by mhvk — COMMENTED** [review]
Looks good generally, but wondering a bit about whether code can be refactored to a common implementation.

**Comment by neutrinoceros** [comment]
Yeah, also the tests are very redundant to satisfy completion meta tests. I'm genuinely wondering wether it's preferable to try and refactor now or later, when we have a complete set of implementations and we can see the whole picture.

**Comment by neutrinoceros** [comment]
rebased and refactored with `@function_helper`, following @mhvk's pattern

**Review by mhvk — APPROVED** [review]
Thanks, looks all good! (I think I'd have used `**kwargs` to combine some, but, really, this is better!)

File: astropy/units/quantity_helper/function_helpers.py [hunk 2]
```diff
@@ -531,6 +532,45 @@ def full_impl(shape, fill_value, *args, **kwargs):
...
+    if out_unit is not UNIT_FROM_LIKE_ARG:
+        a = _as_quantity(a).value
+    return (a, dtype, order), {}, out_unit, None
+
+@function_helper(helps=asarray_impl_2_helps)
+def asarray_impl_2(a, dtype=None, order=None, *, device=None, copy=None):
+    out_unit = getattr(a, "unit", UNIT_FROM_LIKE_ARG)
+    if out_unit is not UNIT_FROM_LIKE_ARG:
+        a = _as_quantity(a).value
+    return (a, dtype, order), {"device": device, "copy":copy}, out_unit, None
```
Comment (neutrinoceros) [inline astropy/units/quantity_helper/function_helpers.py:565]: I don't know how ruff isn't complaining here but let's use this as a pretext to retrigger CI, seeing that automerge seems blocked

```suggestion
    return (a, dtype, order), {"device": device, "copy": copy}, out_unit, None
```

Comment (neutrinoceros) [reply astropy/units/quantity_helper/function_helpers.py:565]: Nevermind the merger happened already.

Comment (mhvk) [reply astropy/units/quantity_helper/function_helpers.py:565]: Weird indeed that it wasn't caught. We can do it elsewhere...

**Pull request merged**